### DISCUSSION TITLE
Arduino nano 33 ble how to flash for opendeck usage

### DISCUSSION THREAD

**Discussion by ratracerecord** [shanteacontrols/OpenDeck#168] [Q&A] [answered]
Hello,

First thank you for all this nice work is Opendeck!
Second i'm a big noob, only sound tech who's trying to build a custom midi controler, so some instruction are a little "blured" for me.

I just bought an arduino nano 33 ble to begin to test some midi controler (in the futur if i achevied to instal properly opendeck on it i would use it with opendeck3 board for more in and out).

So i would like to flash the arduino nano 33 ble with the nano33ble.bin provide here ( https://github.com/shanteacontrols/OpenDeck/releases ) but it seem I can't use the nrf connect app because this have an arduino bootloader.

i try to flash with this type avec cmd line : "C:\Users\ratra\AppData\Local\Arduino15\packages\arduino\tools\bossac\1.9.1-arduino2/bossac.exe" -d --port=COM4 -U -i -e -w "C:\Users\ratra\Downloads/nano33ble.bin" -R 

but no controler detected on the opendeck controle page.

So how could i do to make this arduino nano 33 ble work ? 🗡️ 

Thank's in advance,

Best regards,

Jules.

**Comment by paradajz** [comment]
Hi,

I'm glad you like the project.

Writing clear and easy-to-follow documentation takes a lot of time. The documentation on how to flash Nano 33 BLE will come at some point, but I don't know when. Flashing that particular board is a bit involved and is not that simple, unfortunately. You cannot flash OpenDeck firmware on it with Arduino bootloader - it needs to be flashed with external programmer such as JLink, Black Magic Probe or nRF52840 development kit. Nano has SWD connections on the bottom of the board:

![image](https://user-images.githubusercontent.com/2544094/185927275-eb2ee05f-cdc4-4eb1-921a-a51be61c3766.png)

You need to connect SDWIO, SWDCLK and GND connections to external programmer and then flash it. Fair bit of warning, though: once you flash the board this way, it can no longer be used with Arduino IDE unless you restore the Arduino bootloader on it.

**Comment by ratracerecord** [comment]
Thank you for the quick answer!

So if i understand well i could use nrf52840DK as programmer (using the SEGGER J-Link OB Debugger with debug out functionality) connecting the good pin to SWD, SWCLK and GND from the nrf52840DK P19 Debug Header.

here (https://www.pic-control.com/nrf52840-dk-ble-bluetooth-development-kit-tutorial-start-up-guide/) i can read i could use the NRF connect software to program my nano 33 ble.
In the "methode 4" it say "If there is a connecting board, the nRF52840-DK will program the external board instead of the nRF52840 chip onboard the nRF52840-DK board."

so when the 2 board are connected for program (like say before) i juste have to follow the flahing nrf boards (https://github.com/shanteacontrols/OpenDeck/wiki/Flashing-nRF-boards) and till the nrf52840DK detect the target board on VTG pin, the target board will be the flashed one with the *.hex file.

Am I right, or do i miss something ?

**Comment by paradajz** [comment]
You are right, that is the procedure, yes.

**Comment by ratracerecord** [comment]
Big thank's! :)

**Comment by ratracerecord** [comment, marked as answer]
So just for beeing sure i understand all nicely i had to : 

-connect SWD, SWCLK and GND between booth boards (nrf52840DK as programmer board and nano 33 ble as target board)

-short VTG pin and VDD nrf pin on the nrf52840DK : it gonna tell the nrf52840dk to program the board connected to debug pins and not the nrf52840DK itself.

-connect usb on both board (on the nrf usb port for the nrf52840DK for power and data from the computer; on the usb port of the nano 33 ble only for power)

EDIT : -before flashing the *.hex from nrf connect do i need to put the nano 33 ble in bootloader mode (pushing the buton on it to get the orange pulsating led) ?

-follow the flashing nrf boards page (https://github.com/shanteacontrols/OpenDeck/wiki/Flashing-nRF-boards)


i had some doubt about powering both board separetly (or do i need to power the nano 33 ble from the nrf 52840DK ?) and about the usb connection needed by nrf52840DK (nrf usb or mcu usb ?)...

**Reply by paradajz** [reply]
> EDIT : -before flashing the *.hex from nrf connect do i need to put the nano 33 ble in bootloader mode (pushing the buton on it to get the orange pulsating led) ?

No, there's no need.

> i had some doubt about powering both board separetly (or do i need to power the nano 33 ble from the nrf 52840DK ?) and about the usb connection needed by nrf52840DK (nrf usb or mcu usb ?)...

Both boards need to be powered (programmer/DK and target/Nano). Doesn't matter how (for Nano at least).

**Reply by ratracerecord** [reply]
thank you, i ordered an nrf52840DK, now waiting for it :)

And do i had to connect the programmer/DK via nrf usb or mcu usb ?

**Reply by paradajz** [reply]
> via nrf usb or mcu usb ?

I don't understand the question.

**Reply by ratracerecord** [reply]
on the nrf52840 there is 2 usb port (J2 and J3 here call nrf usb https://infocenter.nordicsemi.com/topic/ug_nrf52840_dk/UG/dk/images/nRF52840/pca10056_debug_output_connector.svg )

**Reply by paradajz** [reply]
Ah. Connect just the debugger.

**Reply by ratracerecord** [reply]
so J2 i suposed.

sorry for being so bad at understanding all of this and again thank you for all your help!

**Comment by ratracerecord** [comment]
Worked perfectly! thank you!

**Reply by FractalFM3** [reply]
Hello there,

thank you so much for your thread and @paradajz support. I had same problem. I will follow your procedure and will try to flash my board.
Question for you: did you used P19 or P20 to connect and if used P19, what pins? I did not find pin description of P19 for SWD, SWCLK and GND.
I think I will use P20 as pins are clearly identified, but just wanted to make sure
Thanks

**Reply by FractalFM3** [reply]
My PCB version is 3.0.0 as per below so there is not VTG on P20 and no GND DETECT as shown in the manual listed in this link (https://www.pic-control.com/nrf52840-dk-ble-bluetooth-development-kit-tutorial-start-up-guide/)
my PCB 3.0.0 looks like this: 
[PCA10056_Schematic_And_PCB.pdf](https://github.com/shanteacontrols/OpenDeck/files/9425817/PCA10056_Schematic_And_PCB.pdf)
<img width="1169" alt="Capture d’écran 2022-08-25 à 16 51 21" src="https://user-images.githubusercontent.com/112095984/186699126-ac4564d4-d997-45e5-bae6-84b4b3e254f8.png">
so What pin on the DK should I use to connect the Nano 33 BLE?
Thanks

**Reply by FractalFM3** [reply]
Finally I succeeded, using same pins a V1.0.0 PCB despite they have no name on P20.
However on Mac several time it has shows JLINK in the nRF Connect and said not recognized device, but after trying a 4th Time then the device showed and I was able to flash Open deck, 

All good Thanks again

**Comment by FractalFM3** [comment]
Hello again,

I did FW update using OpenDeck configurator (nano33BLE.sys) and it went well up to the pop up box saying FW update is finished.
I pushed the reset button (also try after to disconnect the board) but when I go to configurator it does not come up to configuration and always go to UPDATE FW page asking me to load a file like if the Nano was stuck in Bootloader mode. 

During update led is blinking, after reset it is permanently lighting. I rebooted my Mac but always the same. I done several time the update and everytime it was working nicely.

 Just the web configurator takes me directly to update FW. I loaded latest FW (the one he found when I checked for update via the configurator). 

Sorry for the long post, any help appreciated and sorry for being such a newbie

**Reply by paradajz** [reply]
I don't know. Try flashing the v7.3.3 hex file.

**Comment by FractalFM3** [comment]
Thanks for your answer

I downloaded Brave browser and it worked so I suspect it comes from my chrome browser 
Now working nice

**Comment by paradajz** [comment]
Official instructions are now available here:
https://github.com/shanteacontrols/OpenDeck/wiki/Flashing-nRF-boards
